comit/src/order.rs

```diff
@@ -89,27 +84,93 @@ impl BtcDaiOrder {
             created_at: OffsetDateTime::now_utc(),
         }
     }
+}
+
+/// A newtype representing a quantity in a certain base currency B.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Quantity<B> {
+    value: B,
+}
+
+impl Quantity<Bitcoin> {
+    pub fn new(value: Bitcoin) -> Self {
+        Self { value }
+    }
+
+    pub fn sats(&self) -> u64 {
+        self.value.as_sat()
+    }
+
+    /// The [`Bitcoin`] type encapsulates sats and btc well, hence we can just
+    /// provide access to the inner value here.
+    pub fn to_inner(&self) -> Bitcoin {
+        self.value
+    }
+}
+
+#[cfg(test)]
+pub fn btc(btc: f64) -> Quantity<Bitcoin> {
+    Quantity::new(Bitcoin::from_sat(
+        bitcoin::Amount::from_btc(btc).unwrap().as_sat(),
+    ))
+}
+
+/// A newtype representing the price of one unit of the base currency B in the
+/// quote currency Q.
+///
+/// The core idea around of this type is to enforce the unit of the rate when
+/// doing calculations. We achieve that by adding "loud" constructors and
+/// accessors for combinations of base and quote currency.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Price<B, Q> {
+    value: Q,
+    _base: PhantomData<B>,
+}
 
-    /// Returns the price of this order in the given denomination.
+impl Price<Bitcoin, Erc20Quantity> {
+    /// Constructs a new instance of Price where the rate is given in WEI/SAT.
     ///
-    /// The field `price` holds the value in WEI per SATOSHI, hence we need to
-    /// multiply it by 100_000_000 to get WEI per BTC.
-    pub fn price(&self, denom: Denomination) -> Erc20Quantity {
-        let price = self.price.clone();
-        match denom {
-            Denomination::WeiPerSat => price,
-            Denomination::WeiPerBtc => 100_000_000 * price,
+    /// This is how we store the data internally and hence we don't need to do
+    /// any conversions.
+    pub fn from_wei_per_sat(rate: Erc20Quantity) -> Self {
```

Comment:
I still think we should use a a u64 rather than u256. The issue is highly unlikely to arise but going with a u64 prevents us from creating values greater than what u256 supports when doing price * quantity calculations.

Reply:
Max value of a u64:

((2**64) - 1) == 18446744073709551615

= 1.8 x 10e19

Price of 13000 DAI in wei/sat is 

= 13000 x 10e18 / 10e8
= 1.3 x 10e12

So it seems there is enough room in a u64.

I don't see any reason to use a u64 though, since this is DAI why use any type other than what we use for wei, if we did so it would need an explanation.

Reply:
That would limit us to a price of 1.8 billion DAI/BTC. Surely Bitcoin is gonna crack that in the next couple of months. /s

I am not sure to be honest, what is everyone else thinking? @comit-network/comit-devs